torchx/schedulers/kueue_job_scheduler.py

```diff
@@ -0,0 +1,809 @@
+
+    def _submit_dryrun(
+        self, app: AppDef, cfg: KueueJobOpts
+    ) -> AppDryRunInfo[KueueJob]:
+        # map any local images to the remote image
+        images_to_push = self.dryrun_push_images(app, cast(Mapping[str, CfgVal], cfg))
+
+        service_account = cfg.get("service_account")
+        assert service_account is None or isinstance(
+            service_account, str
+        ), "service_account must be a str"
+
+        local_kueue = cfg.get("local_kueue")
+        assert local_kueue is None or isinstance(
+            local_kueue, str
+        ), "local_kueue must be a str"
+        
+        kueue_priority_class = cfg.get("kueue_priority_class")
+        assert kueue_priority_class is None or isinstance(
+            kueue_priority_class, str
+        ), "kueue_priority_class must be a str"
+        
+        resource = app_to_resource(app, service_account, local_kueue,  kueue_priority_class)
+        req = KueueJob(
+            resource=resource,
+            images_to_push=images_to_push,
+        )
+        return AppDryRunInfo(req, repr)
+
+    def _validate(self, app: AppDef, scheduler: str) -> None:
+        # Skip validation step
+        pass
+
+    def _cancel_existing(self, app_id: str) -> None:
+        from kubernetes import client
+        namespace, name = app_id.split(":")
+
+        self._batchv1_api().delete_namespaced_job(
+            namespace=namespace,
+            name=name,
+            body=client.V1DeleteOptions(
+                    propagation_policy='Foreground',
+                    grace_period_seconds=5))
+
+    def _run_opts(self) -> runopts:
+        opts = runopts()
+        opts.add(
+            "namespace",
+            type_=str,
+            help="Kubernetes namespace to schedule Job in",
+            default="default",
+        )
+        opts.add(
+            "service_account",
+            type_=str,
+            help="The service account name to set on the pod specs",
+        )
+        opts.add(
+            "local_kueue",
+            type_=str,
+            help="The Local Kueue name to set on the local Kueue label",
+        )
+        opts.add(
+            "kueue_priority_class",
+            type_=str,
+            help="The kueue priority class name to use for the priority class label",
+        )
+        return opts
+
+    def describe(self, app_id: str) -> Optional[DescribeAppResponse]:
+        from kubernetes import client
+        
+        namespace, name = app_id.split(":")
+        roles = {}
+        roles_statuses = {}
+
+        try:
+            api_instance = self._batchv1_api()
+            job = api_instance.read_namespaced_job_status(name, namespace)
+        except client.ApiException as e:
+            return f"Exception: {e}"
+        
+
+        try: 
+            status = job.status
+        except:
+            status = None
+        
+        if status:
+            condition = status.conditions[0]
```

Comment:
Is the first element of conditions special somehow?

Reply:
It contains the `Suspended` condition that is applied by Kueue.
```
'conditions': [{'last_probe_time': datetime.datetime(2024, 2, 15, 12, 3, 38, tzinfo=tzutc()),
                 'last_transition_time': datetime.datetime(2024, 2, 15, 12, 3, 38, tzinfo=tzutc()),
                 'message': 'Job resumed',
                 'reason': 'JobResumed',
                 'status': 'False',
                 'type': 'Suspended'},
                {'last_probe_time': datetime.datetime(2024, 2, 15, 12, 6, 23, tzinfo=tzutc()),
                 'last_transition_time': datetime.datetime(2024, 2, 15, 12, 6, 23, tzinfo=tzutc()),
                 'message': None,
                 'reason': None,
                 'status': 'True',
                 'type': 'Complete'}],
```